### PR TITLE
Support explicit API mode for complex map keys

### LINKED ISSUE #735: Annotation with `@MapKey(unwrapValue = false)` ("complex map key") doesn't support explicit API mode
The source that `MapKeyCreatorGenerator` generates is missing an explicit return type, giving this error message:
```
Return type must be specified in explicit API mode
```

I'm interested in contributing a fix (along with additional tests) and will be raising a PR in a few minutes. 😃

### PR DESCRIPTION
**Pull request by bddckr**
Resolves #735

### CODE REVIEW

File: integration-tests/library/build.gradle [hunk 1]
```diff
@@ -12,6 +12,10 @@ if (libs.versions.config.generateDaggerFactoriesWithAnvil.get().toBoolean()) {
 
   dependencies {
     kapt libs.dagger2.compiler
+
+    // Necessary because this is what dagger uses when it runs to support instantiating annotations at runtime
+    implementation libs.auto.value.annotations
```
Comment (ZacSweers) [inline integration-tests/library/build.gradle:17]: I'm curious why this is needed now and not before?

Comment (bddckr) [reply integration-tests/library/build.gradle:17]: tl;dr: [Dagger only creates the annotation creator class for the map key when `@MapKey(unwrapValue = false)` is used.](https://github.com/google/dagger/blob/2067f51a476b1f60c3aa94812af0f8f0db661521/java/dagger/internal/codegen/processingstep/MapKeyProcessingStep.java#L68-L69)

As seen in that link, when `unwrapValue = true`, it uses `UnwrappedMapKeyGenerator` instead, which doesn't use `@AutoAnnotation`.

---

Just to clarify:

Without
```kotlin
implementation libs.auto.value.annotations
kapt libs.auto.value.processor
```

we get
```
> Task :integration-tests:library:kaptKotlin
C:\Projects\anvil\integration-tests\library\build\tmp\kapt3\stubs\main\com\squareup\anvil\test\WrappedBindingKey.java:6: error: @AutoAnnotation is a necessary dependency if @MapKey(unwrapValue = false). Add a dependency for the annotation, "com.google.auto.value:auto-value-annotations:<current version>", and the annotation processor, "com.google.auto.value:auto-value:<current version>"
public abstract @interface WrappedBindingKey {
                 ^
```

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MapKeyCreatorGenerator.kt [hunk 2]
```diff
@@ -178,6 +178,7 @@ internal class MapKeyCreatorGenerator : PrivateCodeGenerator() {
         className,
         properties.entries.map { it.value.callExpression }.joinToCode()
       )
+      .returns(className)
```
Comment (JoelWilcox) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MapKeyCreatorGenerator.kt:181]: Nice, coincidentally this is also needed for upgrading to newer KotlinPoet versions

File: integration-tests/library/build.gradle (hunk 1 above)
Comment (JoelWilcox) [reply integration-tests/library/build.gradle:17]: Thanks for sharing the extra context :+1:

**Review by JoelWilcox — APPROVED** [review]

**Pull request merged**